cross-doc-explainer.md

```diff
@@ -0,0 +1,114 @@
+In the future we could examine exposing them in some way to same-site or cross-origin navigations.
+
+## Opt-in
+
+To enable cross-document transitions, both the old and new document need to add a meta tag. This tag
+indicates that the author wants to enable transitions for same-origin navigations to/from this
+Document.
+
+
+### Issues
+
+1. The meta tag can be used to opt-in to other navigation types going forward: same-document, same-site, etc.
+
+1. Using meta-tags prevents the declaration being controlled by media queries, which feels important for `prefers-reduced-motion`.
+
+## Lifecycle
+
+![Lifecycle chart for cross-document transitions](media/mpa-chart.svg)
+
+### Capturing the old state
+
+The old state is captured right before the old document is hidden and a new one is shown.
+In the HTML spec this is defined [here](https://html.spec.whatwg.org/#populating-a-session-history-entry:loading-a-document).
+This can either happen during normal navigations, when the new document is about to be created,
+in BFCache navigations, or when activating a prerendered document.
+
+Before creating the new document (or activating a cached/prerendered one), the UA would update the rendering and snapshot the old document, in the same manner a document is snapshotted for a same-document navigation.
+
+Note that currently there are no planned new events for the exit transition.
+The developer can use existing events like `navigate` or `click` to customize the
+exit transition at moments that should be late enough.
+
+#### Issues
+
+1. Should we enable skipping the transition on exit? This could be done on`pagehide`.
+
+### Capturing the new state
+
+The new state is captured at the first [rendering opportunity](https://html.spec.whatwg.org/#rendering-opportunity)
+of the new document. This allows the new document to use the
+[render-blocking mechanism](https://html.spec.whatwg.org/#render-blocking-mechanism) as a way to
+delay the transition.
+
+For most cases, opting in via the meta tag, styling with the pseudo-elements, and delaying the transition using the
+[render-blocking mechanism](https://html.spec.whatwg.org/#render-blocking-mechanism) should be sufficient. But there are certain cases where further customization is desired, for example:
+
+* Abort the transition if certain conditions apply.
+
+* Modify the pseudo-element style based on certain conditions, e.g. apply transition
+   names only to images that are already loaded.
+
+* Prepare the exit transition right before capture.
+
+There are several options as to how to enable this:
+
+1. Fire an event right before the new document capture. This would be either right
+   before the first `requestAnimationFrame` callback, or after `pageshow` in the reactivation case. Potential name: `reveal`. Note that in the latter case, it's too late to delay the transition via the render-blocking mechanism. This event doesn't
```

Comment:
"reveal" and "pageshow" sound very similar in naming tbh, which makes me think that pageshow might be the right event? maybe?

Reply:
Yea also `visibilitychange`. The problem with `pageshow` is that in the case of a new document, it will wait until the `load` event. So we could use it for reactivate and use the first rAF callback for new document, but that means that developers would have to be extra careful and always use those two callbacks. Another option is that we have a `document.currentTransition` or some such, and `requestAnimationFrame` is *always* the place to handle it.

Reply:
I see. Yeah that's why I think the diagram you have is useful: the relative timing of everything here is hard

I'm personally not a fan of document.currentTransition, since it feels like we still need to define when that object is set and what is its lifetime. For example, in prerender it wouldn't exist always, so at some point we'd populate the currentTransition. It feels like the time at which we would populate it is also the time where we could fire an event. This isn't an opinion that I hold strongly though